mechanics/tools/gdocs-migrate/forest.go

```diff
@@ -0,0 +1,64 @@
+package main
+
+type File struct {
+	id       string
+	name     string
+	parentID string
+	isFolder bool
+}
+
+type Forest struct {
+	fileByID map[string]File
+	children map[string][]File
+	leaves   []File
+}
+
+func NewForest() Forest {
+	return Forest{
+		fileByID: make(map[string]File),
+		children: make(map[string][]File),
+	}
+}
+
+func (fo *Forest) Add(f File) {
+	fo.fileByID[f.id] = f
+
+	fo.children[f.parentID] = append(fo.children[f.parentID], f)
+	if !f.isFolder {
+		fo.leaves = append(fo.leaves, f)
+	}
+}
+
+func (fo *Forest) GetFiles() []File {
+	return fo.leaves
+}
+
+func (fo *Forest) Size() int {
+	return len(fo.fileByID)
+}
+
+func (fo *Forest) FolderCount() int {
+	return fo.Size() - len(fo.GetFiles())
+}
+
+// GetPath returns all the ancestor of a given File in decreasing
+// seniority.
+//
+// This should only be called after the whole forest has ben loaded,
+// i.e. Add(...) has been called for all the Files.
+func (fo *Forest) GetPath(f File) []string {
```

Comment:
This can also be cached if in the struct if it's going to be called more than once.

Reply:
Sure thing, if it is going to be a bottle neck I'll do that.